target/linux/ramips/dts/DWR-512-B.dts

```diff
@@ -0,0 +1,115 @@
+/dts-v1/;
+
+/include/ "rt5350.dtsi"
+
+/ {
+	compatible = "ralink,rt5350-soc";
+	model = "D-Link DWR-512 B";
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		wps {
+			label = "wps";
+			gpios = <&gpio0 0 1>;
+			linux,code = <0x211>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		sms {
+			label = "dwr-512-b:green:sms";
+			gpios = <&gpio0 8 1>;
+		};
+		status {
+			label = "dwr-512-b:green:status";
+			gpios = <&gpio0 9 1>;
+		};
+		2g {
+			label = "dwr-512-b:green:2g";
+			gpios = <&gpio0 17 1>;
+		};
+		3g {
+			label = "dwr-512-b:green:3g";
+			gpios = <&gpio0 19 1>;
+		};
+		sstrengthr {
+			label = "dwr-512-b:red:sigstrength";
+			gpios = <&gpio0 20 1>;
+		};
+		sstrengthg {
+			label = "dwr-512-b:green:sigstrength";
+			gpios = <&gpio0 21 1>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+ 
+		modem3g_enable {
+			gpio-export,name = "modem3g_enable";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 18 1>;
+		};
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	mx25l6405d@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "macronix,mx25l6405d", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <30000000>;
+		fast-read;
+
+                partition@0 {
+                        label = "u-boot";
+                        reg = <0x0 0x30000>;
+                        read-only;
+                };
+
+                partition@30000 {
+                        label = "u-boot-env";
+                        reg = <0x30000 0x10000>;
+                        read-only;
+                };
+
+                factory: partition@40000 {
+                        label = "factory";
+                        reg = <0x40000 0x10000>;
+                        read-only;
+                };
+
+                partition@50000 {
+                        label = "firmware";
+                        reg = <0x50000 0x7a0000>;
+                };
+
+		config: partition@7f0000 {
+			label = "config";
+			reg = <0x7f0000 0x10000>;
+		};
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "jtag", "uartf";
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&config 0xe07a>;
```

Comment:
> I understood that the wmac in the dts point the memory location where the mac address is located.

First of all, the wmac node is just a configuration container for the wireless chip. Have a look at the rt5350.dtsi file, which is included and extended by your dts.

The wireless driver has support for the following device tree properties:

* ralink,eeprom: EEPROM filename that should be used in case the EEPROM needs to be extracted by a userspace programm (target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom)
* ralink,mtd-eeprom: get the EEPROM directly from the named partition at offset 0xn
* mtd-mac-address: get the mac address directly from the named partition at offset 0xn
* mtd-mac-address-increment: increment/decrement the mac address read from flash by n.

You are right, at position 0x4 of the EEPROM is the mac address stored, which will be used by the rt2x00 wireless driver by default. This mac address can be overridden by mtd-mac-address*. But the EEPROM does way more than only including the default mac. The available channels, calibration data and some more hardware settings are stored in the EEPROM.

Your obvious fault is to assume that the ralink,eeprom is related to the MAC address. It is related to the EEPROM which includes the mac address (among others).

Aligning the EEPROM offset parameter of ralink,mtd-eeprom to a position according to a value that looks like the mac address, doesn't mean that you necessarily found the EEPROM.

> 0000e080 2e a0 78 6d 00 f5 00 80 00 02 50 53 00 01 00 50 |..xm......PS...P|
0000e090 18 21 d5 28 ff ff ff ff ff ff ff ff ff ff ff ff |.!.(............|
>
>@0xe08a I can also found the 0x5053 referencing the SoC.

Yeah, that ```50 53 00 01 00 50 18 21 d5 28``` part looks similar to the **start** of other ralink EEPROMs I've seen. But I would expect to see values other than 0xFF afterwards. Since you pasted only the 0x0000e090 line, I assumed everything that follows is 0xFF as well. Might be a wrong assumption and you have just not pasted enough.

Furthermore, the potential mac address at  0xe08a + 0x4 doesn't look like the one it should be.

>Proceeding in this way, as you can see, the lede system is able to get the same mac address:
[ 12.321593] rt2800_wmac 10180000.wmac: loaded eeprom from mtd device "config"
[ 12.335989] ieee80211 phy0: rt2x00_set_rt: Info - RT chipset 5350, rev 0500 detected
[ 12.351537] ieee80211 phy0: rt2x00_set_rf: Info - RF chipset 5350 detected
[ 12.366677] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'

As said earlier, I haven't seen a RT5053 EEPROM yet . It's not unlikely that everything is fine. I was just curious because the dts of the RT5053 boards I've check have the EEPROM at a different position.

It wasn't a request to change anything. I just wanted to make sure that everything is right. I asked for a copy of your flash to check it by myself. In case I'm not sure with something, I don't please other people to do the work for me. So, can I get a dump of the DWR-512 flash *before* you have written/overwritten/remove anything to/from flash. Dropbox link for example is fine. No need to paste the whole 16MB flash file dump here.

> PS: How can you properly format the hexdump in this git conversation?

https://enterprise.github.com/downloads/en/markdown-cheatsheet.pdf topic "fenced code blocks"